transition analysis docs to numpy style

orbeckst opened the pull request:
Use this WIP PR to accumulate doc fixes for analysis. When we have enough we can merge or squash-merge. I didn't want to do a proper PR for a single simple reST fix...

Changes made in this Pull Request:
 - doc fixes in analysis (eg SeeAlso, Example sections, ...)
 - transition analysis [docs to numpy style](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#writing-docstrings)

Note that @jbarnoud has been tackling *all other docs* in PR #1247.


PR Checklist
------------
 - n/a Tests?
 - [x] Docs?
 - n/a CHANGELOG updated?
 - n/a Issue raised/referenced?

orbeckst commented:
I cleaned up some code in `analysis.gnm` and apparently broke things. Will need to look into this.

orbeckst commented:
fixed gnm issue, rebased and force-pushed

orbeckst commented:
... and travis exceeded time so I restarted the job.

orbeckst commented:
Full test exceeded time again. I'll try rolling back whatever small code changes I made in analysis.gnm.

(Locally
```
./mda_nosetests -v ./analysis/test_gnm.py
```
takes 122.4 s on a single 3.1 GHz i7 core with most of the time spent on `test_gnm.TestGNM.test_closeContactGNMAnalysis`.)

orbeckst commented on package/MDAnalysis/analysis/gnm.py:
```diff
@@ -337,12 +353,9 @@ def generate_kirchoff(self):
                 if jcounter > icounter and _dsq(positions[icounter], positions[jcounter]) <= cutoffsq:
                     iresidue, jresidue = residue_index_map[icounter], residue_index_map[jcounter]
-                    if self.MassWeight:
-                        contact = 1.0 / (len(self.ca.residues[iresidue].atoms) * len(self.ca.residues[jresidue].atoms)) ** 0.5
```
This line was responsible for the huge amount of time that the GNM test `TestGNM.test_closeContactGNMAnalysis` took.

kain88-de replied:
Can you push your fix to a new branch separate from this?

kain88-de replied:
also this logic appears also in the other class in this file. It would be nice if you can change it their too.

orbeckst replied:
Will fix elsewhere; apparently only mattered in the instance were I fixed it because many more atoms were used.

I can try to disentangle but their were also some cleanups in gnm in an earlier commit. I will look at a local interactive rebase and see if I can merge these commits.

orbeckst replied:
Er, actually, *where* did you see another instance of this problem? `GNMAnalysis.generate_kirchoff()` does not use custom weights. I am not sure where else this needs fixing.

kain88-de replied:
Looks like I misremembered.  I thought both classes use the custom weights.

orbeckst replied:
See PR #gnm – all analysis.gnm changes were removed from this PR.

orbeckst replied:
@richardjgowers / @dotsdl would you *expect* a lookup such as `len(self.ca.residues[iresidue].atoms` to be slow? In particular, is `self.ca.residues` the bottle neck?

richardjgowers replied:
So assuming `self.ca` is an AtomGroup, 
 - `ca.residues` will [be performing a `np.unique` on all the residue indices in the group](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/groups.py#L1522),
 - `residues[iresidue]` is just slicing a numpy array
 - `residue.atoms` [will be looking up a single entry in a residue->atoms table](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/groups.py#L2556).

So nothing too offensively slow from what I can see, but you might be able to calculate them all ahead of time (as these memberships won't change) if you're looping over these things a lot.

orbeckst commented:
Optimized [a single line](https://github.com/MDAnalysis/mdanalysis/pull/1240/files#r108870359) in `analysis.gnm.closeContactGNMAnalysis.generate_kirchhoff()` and locally reduced the run time of the longest test `TestGNM.test_closeContactGNMAnalysis` by about 4-5 fold. 

(Also rebased and force pushed.)

kain88-de commented on package/MDAnalysis/analysis/gnm.py:
```diff
@@ -314,14 +327,17 @@ def __init__(self, universe, selection='protein', cutoff=4.5, ReportVector=None,
         cutoffsq = self.cutoff ** 2
 
+        # cache sqrt of residue sizes (slow) so that sr[i]*sr[j] == sqrt(r[i]*r[j])
+        sqrt_res_sizes = np.sqrt([r.atoms.n_atoms for r in  self.ca.residues]) if self.MassWeight else None
```
Btw is that really a weights defined by the mass? It looks only like a estimate of the actual mass. Would `size` be a better name?

Since you are refactoring this code already. It be nice if this variable was renamed `weights` and then have options `None` or `'size'`.

orbeckst replied:
It's not really weight but that's what was in the original code. 

I'll rename the variable but leave the kwarg as it is but add a note.

orbeckst replied:
See e335c375963d53e8300ee4ede45de6eea158f999

orbeckst replied:
Done in PR #1272

orbeckst commented:
I moved the GNM improvements to separate PR #1272  and rebased and force-pushed.

orbeckst commented:
Yes, that's what's being done now. Speed up of about 5 times. 

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Mar 30, 2017 um 5:43 schrieb Richard Gowers <notifications@github.com>:
> 
> but you might be able to calculate them all ahead of time (as these memberships won't change) if you're looping over these things a

orbeckst commented:
rebased and force pushed

orbeckst commented:
All analysis docs are now numpy style.

orbeckst commented:
Somone has to review it, otherwise it can't be merged...

orbeckst commented on package/MDAnalysis/analysis/hbonds/hbond_analysis.py:
```diff
@@ -214,8 +214,8 @@ class HydrogenBondAnalysis_OtherFF(HydrogenBondAnalysis):
 .. _`10.1002/prot.340090204`: http://dx.doi.org/10.1002/prot.340090204
 
 
-Example
```
Just fyi: with numpy docs, never use `Example` or `Examples` in a normal section context. sphinx napoleon rewrites it. So you have to ne a bit creative with labelling sections in the main part of the page.

kain88-de commented on package/MDAnalysis/analysis/nuclinfo.py:
```diff
@@ -113,24 +113,31 @@
+
 
-    .. NOTE:: If failure occurs be sure to check the segment identification.
+    .. Note:: If failure occurs be sure to check the segment identification.
```
Why not the numpy notes section

orbeckst replied:
Changed to Notes section (rebased so it now shows up in commit 056f0976100e02c82a354a76be79f0669549926e )

orbeckst commented on package/MDAnalysis/analysis/psa.py:
```diff
@@ -405,6 +463,13 @@ def hausdorff_wavg(P, Q):
      >>> hausdorff_wavg(P,Q[::-1]) # weighted avg hausdorff dist w/ Q reversed
      2.5669644353703447
+
+    Notes
```
Yes, its `Notes` as in https://github.com/numpy/numpy/blob/master/doc/example.py

orbeckst commented on package/MDAnalysis/analysis/nuclinfo.py:
```diff
@@ -146,24 +153,31 @@ def wc_pair(universe, i, bp, seg1="SYSTEM", seg2="SYSTEM"):
 
-    .. NOTE:: If failure occurs be sure to check the segment identification.
+
+    .. note:: If failure occurs be sure to check the segment identification.
```
Changed to Notes section (rebased so it now shows up in commit 056f0976100e02c82a354a76be79f0669549926e )

orbeckst commented:
Incorporated @kain88-de 's suggestions, rebased into the appropriate commit, and rewrote some of the commit messages. I think I am done. 

When @jbarnoud finishes PR #1247 then we will have transitioned *all* our docs to numpy style.

kain88-de commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -943,45 +947,46 @@ def fasta2select(fastafilename, is_aligned=False,
-    :func:`sequence_alignment`, which does not require external
-    programs.
+
+    .. SeeAlso::
```
Any idea how to change those automatically to the old style? My first try `find . -name '*py' -exec sed -i "s/.. SeeAlso::/See Also \n--------\n/" {} \; ` doesn't work. It can't deal with potential indentation of the initial see also paragraph.

jbarnoud replied:
I had a quick look and I ended up with this:

```
sed -re 's/^(( *).. SeeAlso::( *))/\2See Also\n\2--------\n\2/g' package/MDAnalysis/lib/util.py | less
```

You need the `-r` or you will have to escape all the parentheses, and you may not have the `\2` syntax. The `\2` is there to report the indentation.
